### ISSUE TITLE
AWS - Enable setting custom domain instead of hardcoded amazonaws.com

### ISSUE THREAD

**Issue by tkint** [http4k/http4k-connect#17]
Hello!

In our organization, there is an on-premise S3 installation on an internal domain and I noticed that http4k-aws and http4k-connect have `amazonaws.com` domain hardcoded. 

When I change the domain with `ClientFilters.SetHostFrom`, the signature process seems to fail: I get an AccessDenied response, even if the request is correct.

By trying the same request with Postman and its included AWS Signature feature, I got a different SignedHeaders :
 - Postman: `SignedHeaders=host;x-amz-content-sha256;x-amz-date,`
 - Http4k-connect: `SignedHeaders=host;x-amz-date;x-forwarded-host,`

So I think there are those two possibilities, either the signature is wrong due to a late host change, or the signature needs to sign `x-amz-content-sha256` too.

Either way, it would be nice to be able to set the base domain.

Here is the code I tried with:
```kotlin
private fun S3Bucket.Companion.customHttp(
    uri: Uri,
    region: Region,
    bucketName: BucketName,
    credentials: AwsCredentials,
    httpClient: HttpHandler,
) = object : S3Bucket {
    private val http = signAwsRequests(uri, region, bucketName, credentials).then(httpClient)

    private fun signAwsRequests(
        uri: Uri,
        region: Region,
        bucketName: BucketName,
        credentials: AwsCredentials
    ): Filter =
        // The final uri is like: https://<bucket-name>.<internal-domain>/<bucket-key>
        ClientFilters.SetHostFrom(uri.host("${bucketName.value}.${uri.host}"))
            .then(ClientFilters.SetXForwardedHost())
            .then(
                ClientFilters.AwsAuth(
                    scope = AwsCredentialScope(region.value, awsService.value),
                    credentials = credentials,
                    clock = Clock.systemDefaultZone(),
                    payloadMode = Payload.Mode.Signed,
                )
            )

    override fun <R> invoke(action: S3BucketAction<R>): Result<R, RemoteFailure> =
        action.toResult(http(action.toRequest()))
}
```

And the call:
```kotlin
S3Bucket.customHttp(
    uri = Uri.of(url), // url is like: https://<internal-domain>
    region = Region.of(region),
    bucketName = BucketName.of(bucket),
    credentials = AwsCredentials(accessKeyId, accessKeySecret),
    httpClient = httpClient.debug(),
)
```

Thank you for your help :)

**Comment by daviddenton** [comment]
So I've created a Gist with the code that I'm running. The Auth header is based on the preceding Headers - which is why you get the 3 you do. The x-forwarded-host is actually used for the FakeS3 implementation, so you can omit that it you're not using it. 

https://gist.github.com/daviddenton/452d35c462dfb8d99e3f92dd95e416d7

Is this the code that you're running and do you see the same output?

**Comment by tkint** [comment]
Yes, this is the code I am running and I get the same output. I'm using the FakeS3 for dev environement only, so I will let the `X-Forwarded-Host` header for this use case.

**Comment by daviddenton** [comment]
You mentioned "either the signature is wrong due to a late host change,". Just checking that you aren't actually doing this anymore because it's build into your code now? Adding/changing the host later will affect it.

Also - you should get more details from S3 about why the access was denied - probably that the signature didn't match?

**Comment by tkint** [comment]
I am trying to simply get a file:
```kotlin
val s3Bucket = S3Bucket.customHttp(
    uri = Uri.of(url), // url is like: https://<internal-domain>
    region = Region.of(region),
    bucketName = BucketName.of(bucket),
    credentials = AwsCredentials(accessKeyId, accessKeySecret),
    httpClient = httpClient.debug(),
)
s3Bucket[BucketKey.of("<bucket-key>")]
```

This method is working when the visibility of the file is set to public and without the `ClientFilters.AwsAuth`
 filter.

Here are the request and response of a signed call:
```
GET https://<bucket-name>.<internal-domain>/<bucket-key> HTTP/1.1
X-Forwarded-Host: <bucket-name>.<internal-domain>
host: <bucket-name>.<internal-domain>
x-amz-date: 20210210T084405Z
Authorization: AWS4-HMAC-SHA256 Credential=<access-key-id>/20210210/eu-west-3/s3/aws4_request, SignedHeaders=host;x-amz-date;x-forwarded-host, Signature=bf0bd3f39a6525d6768ad971fc0d6d96aeee5f2919e0b8b9a215ba2edc428470
x-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855


***** RESPONSE 403 to GET: https://<bucket-name>.<internal-domain>/<bucket-key> *****
HTTP/1.1 403 Forbidden
connection: keep-alive
content-length: 174
content-type: application/xml
date: Wed, 10 Feb 2021 08:44:05 GMT
server: nginx
x-amz-id-2: 19e83001234daa90f4e7
x-amz-request-id: 19e83001234daa90f4e7

<?xml version="1.0" encoding="UTF-8"?><Error><Code>AccessDenied</Code><Message>Access Denied</Message><Resource></Resource><RequestId>19e83001234daa90f4e7</RequestId></Error>
```

**Comment by daviddenton** [comment]
It's quite difficult for us to know what is going on here as we obviously don't have access to your S3. We have tests to prove that the implementation works ok against public S3.

I wonder if it's the fact that the content-sha header isn't in the signing - this could be a custom config from your worksplace?

It might be worth copying the http4k implementation and seeing if signing WITH that header makes it work. If it does (and we can also replicate the same "it works" behaviour in general S3, we can look at making the change in http4k-aws (which is where the auth code actually lives)

**Comment by tkint** [comment]
Okay, I will try to implement a working solution based on the sources, then communicate it to you :) Thank you for your time and work

**Comment by daviddenton** [comment]
You're welcome - thanks for helping us out to figure out what's going on - not everyone is so helpful!

Aso, we don't tolerate bugs around here.. 😉

**Comment by tkint** [comment]
Okay we finally got the time to do some tests and found a pretty simple solution. We had to add `x-amz-content-sha256` in signed headers.

See the PR on `http4k`: https://github.com/http4k/http4k/pull/592

**Comment by daviddenton** [comment]
@tkint We've now released http4k 4.4.0.1 with that fix in so you should be able to use that version instead of the default that comes with connect.

A release of http4k-connect with the upgraded http4k bundled will be forthcoming 😄 ,

**Comment by daviddenton** [comment]
I'm releasing 2.15.2.0 with the upgraded http4k to fix this.
